odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+        Parameters
+        ----------
+        other : `Functional` or scalar
+            `Functional`:
+            The `Functional.domain` and `Functional.range` of ``other``
+            must match this functional's  `Functional.domain` and
+            `Functional.range`.
+
+            scalar:
+            The scalar needs to be in this functional's `Functional.range`.
+
+        Returns
+        -------
+        add : `Functional`
+            Addition result
+
+            If ``other`` is in ``Functional.range``, ``add`` is a
+            `FunctionalScalarSum`.
+
+            If ``other`` is a `Functional`, ``add`` is a `FunctionalSum`.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return NotImplemented
+
+    def __radd__(self, other):
+        """Return ``other + self``.
+
+        Since addition is commutative, also for functionals, this is similar
+        to `Functional.__add__`.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return NotImplemented
+
+    def __sub__(self, other):
+        """Return ``self - other``."""
+        return self.__add__(-1 * other)
+
+    @property
+    def grad_lipschitz(self):
+        """Lipschitz constant for the gradient of the functional"""
+        return self._grad_lipschitz
+
+
+class FunctionalLeftScalarMult(Functional, OperatorLeftScalarMult):
+
+    """Scalar multiplication of functional from the left.
+
+    Given a functional ``f`` and a scalar ``scalar``, this represents the
+    functional
+
+        ``(scalar * f)(x) = scalar * f(x)``.
+    """
+
+    def __init__(self, func, scalar):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The functional that is to be scaled.
+        scalar : `float`
+            The scalar with which the functional is scaled.
+        """
+
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        scalar = func.range.element(scalar)
+
+        if scalar > 0 or scalar < 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=func.is_linear,
+                                grad_lipschitz=(
+                                    np.abs(scalar) * func.grad_lipschitz))
+
+        elif scalar == 0:
+            Functional.__init__(self, domain=func.domain, linear=True,
+                                grad_lipschitz=0)
+
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
```

Comment:
Remove this, we cast `scalar` to a field element, so we can always compare. Errors are raised earlier if input is garbage.